### ISSUE TITLE
mg_set_request_handler race condition

### ISSUE THREAD

**Issue by hcodina** [civetweb/civetweb#588]
A race condition exist between removing an handler using `mg_set_request_handler`
and a worker thread using this handler.

On the worker thread (handle_request function) : 
``` 
get_request_handler(conn,is_websocket_request,&callback_handler,...);
...
callback_handler(conn, callback_data);
```

Between the call to `get_request_handler` to retreive handler and the handler call `callback_handler(conn, callback_data)`, `mg_set_request_handler` can be called, modifying the handler (callback_data, callback) or simply removing it.
The handler called is in this case obsolete and associated data (callback_data) can be invalid.

A protection should be done in order to avoid modifying an handler between `get_request_handler` and its call.

Not that this will probably imply that 'mg_set_request_handler' should not be called from an handler (deadlock).

A rwlock (I am not sure about rwlock presence on windows) can be used to improve performance in this case and allowing multiple handlers running in parallel.

handle_request : 
```
pthread_rwlock_rdlock(conn->ctx->rwlock);
get_request_handler(conn,is_websocket_request,&callback_handler,...);
...
callback_handler(conn, callback_data);
pthread_rwlock_unlock(conn->ctx->rwlock);
```

mg_set_request_handler :
```
pthread_rwlock_wrlock(conn->ctx->rwlock);
... actual code ...
pthread_rwlock_unlock(conn->ctx->rwlock);
```
And so call mg_set_request_handler from an handler is forbidden.

Hope this could help in civetweb improvements.

**Comment by bel2125** [comment]
This kind of deadlock should be prevented by the `mg_lock_context` call in `get_request_handler` and `mg_set_handler_type`. Did you really see this problem in practice or did you assume it can happen?
If you assumed it from a code review, did you see the `mg_lock_context` call?

I can check the code again, if all relevant code paths have `mg_lock/unlock_context` calls.

Windows has a `SRWLock` call starting from Windows Server 2008.
https://msdn.microsoft.com/en-us/library/windows/desktop/aa904937(v=vs.85).aspx
Up to now CivetWeb was also running on Server 2003, XP and XP professional x64, but all these versions are now out of long term maintenance, so new releases of CivetWeb might actually use this API function.

**Comment by hcodina** [comment]
I see `mg_lock/unlock_context`. They are called inside `get_request_handler` and `mg_set_handler_type` in order to protect handler modification (insertion, removal, update) and retrieval.
This lock does not protect handler "use" (callback call) against handler modification or removal.
Maybe mg_lock/unlock_context must be done outside `get_request_handler` in order to protect retrieval and use against modification (insertion, removal, update).
Something like : 
```
mg_lock_context(conn->ctx);
get_request_handler(conn,is_websocket_request,&callback_handler,...);
...
callback_handler(conn, callback_data);
mg_unlock_context(conn->ctx);
```

As a simple mutex is used in `mg_lock/unlock_context` handling request in parallel will not be done anymore. The use of `rwlock` (posix) or `SRWLock` (windows) will solve parallel handling.

**Comment by bel2125** [comment]
I did not lock the use of handler callback calls against handler list modification, because using a callback may take an indefinite time - seconds, minutes, hours ... even days, e.g. if you are using a websocket connection as a bridge for other protocols.
I don't see any reason why you have to prohibit adding a new handler while some completely different handler is running.

I think the only relevant question is:
Does it harm to remove a handler callback while it is still used?
The immediate consequence is, new calls do not enter this callback anymore.
Tasks currently using this handler will continue (and maybe finish or not).
Without additional operation by the caller, the function does not "disappear", and the user argument does not get invalid.
You can make a function "disappear", if it's in a DLL/so, and you unload this explicitly by a FreeLibrary/dlclose call. You can make the user argument invalid if it's a pointer to some structure and you free this pointer. Before I tell you "just don't do that", please tell me a little more about your use case. If it really requires an R-W lock, I would need one lock per handler, any global lock will yield in starvation if you have just one long or frequently running callback and try to remove a completely different, unused callback.

**Comment by hcodina** [comment]
I fully understand your starvation problem.

My use case is : 
```
struct my_data *my_service_init(...)
{
  /* Init my service */
  data = malloc(...);
  data->resources = malloc(...);
  obj_init(&data->obj);
  /* Ok, everything is initialized -> register callback for civetweb use 
   * my_handler uses data->ressources and data->obj to do its job
   */
  mg_set_request_handler(...,my_handler,"/my_service",data);
  return data;
}

void my_service_exit(struct my_data *data)
{
  /* Unregister handler */
  mg_set_request_handler(...,NULL,"/my_service",NULL);
  /* mg_set_request_handler has unregister my handler -> From here, it cannot be called anymore
   * -> I can release my resources 
   */
  obj_exit(&data->obj);
  free(data->resources);
  free(data);
}

```
In my use case, i need to be sure that the handler i previously registered cannot be called after `mg_set_request_handler(NULL,"/my_service"...,NULL)` returns in order to release associated resources. If the handler can be called, i cannot release my resources just after mg_set_request. In this case, when can i release ? When can i be sure that this handler will not be called anymore ?

To avoid rwlock and its associated starvation, you can have :
- a ref count in order to know if handler will be used by a worker thread
- a pthread_cond_wait to wait the end of use if needed

worker_thread :
- handler found -> ref_count ++
- End of handler use -> ref_count -- + pthread_cond signal

mg_set_request_handler :
- while (ref count) : pthread_cond wait

The ref count and pthread_cond is per handler.

Of course, if you have an idea i can implement in "my_service" (my_service_init, my_service_exit, my_handler) to solve my use case, i'll take it.

**Comment by bel2125** [comment]
... I split it up in two posts ...

not a solution
===

Doing your own refcount has a problem, if the handler task is interrupted between mg_context_unlock in civetweb.c and the atomic_inc in your handler

```C
int my_handler(conn, data)
{
  atomic_inc(data->refcount);
  ...
  atomic_dec(data->refcount);
}

struct my_data *my_service_init(...)
{
  /* Init my service */
  data = malloc(...);
  data->refcount = 0;
  data->resources = malloc(...);
  obj_init(&data->obj);
  /* Ok, everything is initialized -> register callback for civetweb use 
   * my_handler uses data->ressources and data->obj to do its job
   */
  mg_set_request_handler(...,my_handler,"/my_service",data);
  return data;
}


void my_service_exit(struct my_data *data)
{
  /* Unregister handler */
  mg_set_request_handler(...,NULL,"/my_service",NULL);
  /* mg_set_request_handler has unregister my handler -> From here, it cannot be called anymore
   * -> I can release my resources 
   */
  obj_exit(&data->obj);

  do {
    Sleep(100);
  } while (data->refcount > 0);

  free(data->resources);
  free(data);
}
```

You would not have this race condition, if the refcount++ is inside the mg_context_lock in civetweb.c.

**Comment by bel2125** [comment]
might be a solution
===

You could add another indirection

```C

int my_handler(conn, argdata)
{
  data = get_data_from_handle(argdata)
  if (data == NULL) return 0;
  ...
}


struct my_data *my_service_init(...)
{
  /* Init my service */
  data = malloc(...);
  data->resources = malloc(...);
  obj_init(&data->obj);
  /* Ok, everything is initialized -> register callback for civetweb use 
   * my_handler uses data->ressources and data->obj to do its job
   */
  argdata = create_handle_for_data(data);
  mg_set_request_handler(...,my_handler,"/my_service", argdata);
  return argdata;
}


void my_service_exit(struct my_data *argdata)
{
  /* Unregister handler */
  data =  remove_handle_for_data(argdata);
  mg_set_request_handler(...,NULL,"/my_service",NULL);
  /* mg_set_request_handler has unregister my handler -> From here, it cannot be called anymore
   * -> I can release my resources 
   */
  obj_exit(&data->obj);

  free(data->resources);
  free(data);
}
```

Here `create_handle_for_data`, `get_data_from_handle` and `remove_handle_for_data` act on a static data structure such as a singleton map (C++) or a static array (C). If you add a reader lock on top of  my_handler and a writer lock on top of my_service_exit, you solved it outside civetweb.c.

**Comment by hcodina** [comment]
> Doing your own refcount has a problem, if the handler task is interrupted between mg_context_unlock in civetweb.c and the atomic_inc in your handler

I known that. I can not have a reliable solution without the help of civetweb internals.

`create_handle_for_data, get_data_from_handle and remove_handle_for_data` way : 
As i will never be sure that my_handler will never be called after mg_set_request_handler(...,NULL,"/my_service",NULL); i need to keep minimal resources needed by get_data_from_handle available. The lock on global_handler_lock cannot prevent the call to my_handler from civetweb. This mutex cannot be freed. 
Whatever i do, i have to deal with the fact that my_handler can be called after my_service_exit returns.
I have no way to protect against this.

Civetweb can, on the other hand, help to avoid the call to my_handler after i unregister it.
The refcount and the pthread_cond_signal/wait must be implemented inside civetweb.

> worker_thread : 
handler found -> ref_count ++
End of handler use -> ref_count -- + pthread_cond signal

ref_count++ is implemented in `get_request_handler`
ref_count -- + pthread_cond_signal is implemented in `handle_request` just after `callback_handler(conn, callback_data);`
ref_count use : Avoid starvation. 
pthread_cond_signal use : Synchronize with mg_set_request_handler

> mg_set_request_handler : 
while (ref count) { pthread_cond wait; }

The `while (ref_count) { pthread_cond wait; }` is implemented in `mg_set_request_handler`
Its aim : Synchro with my_handler get and call from worker_thread -> Guarantee that it will never called after mg_set_request_handler` returns.

**Comment by bel2125** [comment]
> I can not have a reliable solution without the help of civetweb internals.

I think using `get_data_from_handle` (could be an `std::map<void*, t_your_struct*>[argdata]` or some C implementation) with the rwlock in your handler could work without changing civetweb internals.

This does not mean, an implementation in the civetweb internals would not be beneficial.

Did you already prepare something you could share with a pull request?

**Comment by hcodina** [comment]
I have not anything to share yet but i will work on it. I will let you know when i am done. I am quite busy but i will do my best as soon as possible.

**Comment by bel2125** [comment]
Solved by PR  #589
